### ISSUE TITLE
Should `copy` be part of the optional interface for `AbstractArray`?

### ISSUE THREAD

**Issue by thchr** [JuliaLang/julia#41918]
According to the [AbstractArray interfaces documentation](https://docs.julialang.org/en/v1.8-dev/manual/interfaces/#man-interface-array) it is not necessary to extend `copy` for an `AbstractArray` subtype. Instead, one should extend `similar(A, ::Type{S}, dims::Dims)`. This advice works well for mutable arrays. As an example, defining a mutable array wrapper:
```jl
import Base: size, IndexStyle, getindex, setindex!, similar

struct Wrapper <: AbstractMatrix{Int}
    data::Matrix{Int}
end
size(w::Wrapper) = size(w.data)
IndexStyle(::Wrapper) = IndexLinear()
getindex(w::Wrapper, i::Int) = getindex(w.data, i)
setindex!(w::Wrapper, v, i::Int) = setindex!(w.data, v, i)
similar(w::Wrapper, ::Type{Int}, dims::Dims{2}) = Wrapper(similar(w.data, Int, dims))

w = Wrapper([1 2 3; 4 5 6; 7 8 9])
```
Then, doing `copy(w)` will return an `MWrapper` as one might hope it would.

However, if we want to implement an ("pretend") immutable array by instead letting
```jl
setindex!(w::Wrapper, v, i::Int) = error("let's pretend this is an immutable array...")
```
then we of course get an error from `copy(w)` due to `setindex!`/`similar` interaction. We might try to fix that by _not_ defining `similar` for `Wrapper` (i.e. relying on the method fallback), in which case `copy(w)` returns a `Matrix{Int}` which is also not great [[1]](#footnote1).
Accordingly, it seems one must explicitly extend `Base.copy(w::Wrapper) = Wrapper(copy(w.data))` in this case.

**TLDR:** Should the `AbstractArray` interfaces manual section include a line akin to:

> If you are defining an immutable `AbstractArray` type (i.e., with undefined `setindex!`), you should extend `copy` as well as `similar`.

---

<a name="footnote1">[1]</a> One way in which this is not great is if I e.g. also extend `*` for `Wrapper`:
```jl
import Base: *
(*)(w1::Wrapper, w2::Wrapper) = Wrapper(w1.data * w2.data)
```
Now, since `copy` no longer returns a `Wrapper`, I've introduced a type-instability in `Base.power_by_squaring(w::Wrapper, p::Int)` (it returns `copy(w)` for `p=1` but uses `*` for `p=2,3,...`; so return type is now `Union{Matrix{Int}, Wrapper}`). So now `^` is type-unstable for literal numbers (I suppose const-propagation might take care of that but still...).

**Comment by mbauman** [comment]
The "true" optional methods to implement for `AbstractMatrix` is... long.  Varying array types might want to extend any one of `methodswith(AbstractMatrix; supertypes=true)` — some 650 methods.

It'd probably be better to have specific guidance for immutable arrays, incorporating some of the learnings from StaticArrays.jl.
